### PR TITLE
Implement `--use-loop-variable-after-loop`

### LINKED ISSUE #2067: [SciPy][Specfun]: Divergence in test output `test_riccati_yn`
This error is because of the below given subroutine in `scipy/special/specfun`, I am trying to find a minimal way to reproduce this error.
```fortran
        SUBROUTINE RCTY(N,X,NM,RY,DY)
C
C       ========================================================
C       Purpose: Compute Riccati-Bessel functions of the second
C                kind and their derivatives
C       Input:   x --- Argument of Riccati-Bessel function
C                n --- Order of yn(x)
C       Output:  RY(n) --- x·yn(x)
C                DY(n) --- [x·yn(x)]'
C                NM --- Highest order computed
C       ========================================================
C
```

```console
% python dev.py test -t scipy.special.tests.test_basic --verbose
💻  ninja -C /Users/pgoswami/Desktop/scipy/build -j8
ninja: Entering directory `/Users/pgoswami/Desktop/scipy/build'
[4/4] Generating scipy/generate-version with a custom command
Build OK
💻  meson install -C build --only-changed
Installing, see meson-install.log...
Installation OK
SciPy from development installed path at: /Users/pgoswami/Desktop/scipy/build-install/lib/python3.10/site-packages
Running tests for scipy version:1.12.0.dev0+1345.9b3bbc1, installed at:/Users/pgoswami/Desktop/scipy/build-install/lib/python3.10/site-packages/scipy
================================================ test session starts =================================================
platform darwin -- Python 3.10.12, pytest-7.4.0, pluggy-1.2.0 -- /Users/pgoswami/mambaforge/envs/scipy-dev/bin/python
cachedir: .pytest_cache
rootdir: /Users/pgoswami/Desktop/scipy
configfile: pytest.ini
plugins: timeout-2.1.0, cov-4.1.0, xdist-3.3.1
collected 1 item                                                                                                     

scipy/special/tests/test_basic.py::TestRiccati::test_riccati_yn FAILED                                         [100%]

====================================================== FAILURES ======================================================
____________________________________________ TestRiccati.test_riccati_yn _____________________________________________
scipy/special/tests/test_basic.py:3754: in test_riccati_yn
    assert_array_almost_equal(C, special.riccati_yn(n, x), 8)
        C          = array([[-0.98006658, -5.09900222],
       [ 0.19866933, 24.51494452]])
        N          = 2
        n          = 1
        self       = <scipy.special.tests.test_basic.TestRiccati object at 0x107a12170>
        x          = 0.2
        y          = -25.495011100006344
        yp         = 250.04977811085723
/Users/pgoswami/mambaforge/envs/scipy-dev/lib/python3.10/contextlib.py:79: in inner
    return func(*args, **kwds)
        args       = (array([[-0.98006658, -5.09900222],
       [ 0.19866933, 24.51494452]]), (array([-0.98006658, -5.09900222]), array([0.19866933, 0.        ])), 8)
        func       = <function assert_array_almost_equal at 0x1079bb7f0>
        kwds       = {}
        self       = <contextlib._GeneratorContextManager object at 0x1079c2680>
/Users/pgoswami/mambaforge/envs/scipy-dev/lib/python3.10/contextlib.py:79: in inner
    return func(*args, **kwds)
E   AssertionError: 
E   Arrays are not almost equal to 8 decimals
E   
E   Mismatched elements: 1 / 4 (25%)
E   Max absolute difference: 24.51494452
E   Max relative difference: 5.58829599e-16
E    x: array([[-0.98006658, -5.09900222],
E          [ 0.19866933, 24.51494452]])
E    y: array([[-0.98006658, -5.09900222],
E          [ 0.19866933,  0.        ]])
        args       = (<function assert_array_almost_equal.<locals>.compare at 0x1069fd900>, array([[-0.98006658, -5.09900222],
       [ 0.19866933, 24.51494452]]), (array([-0.98006658, -5.09900222]), array([0.19866933, 0.        ])))
        func       = <function assert_array_compare at 0x1079bb640>
        kwds       = {'err_msg': '', 'header': 'Arrays are not almost equal to 8 decimals', 'precision': 8, 'verbose': True}
        self       = <contextlib._GeneratorContextManager object at 0x1079c25f0>
============================================== short test summary info ===============================================
FAILED scipy/special/tests/test_basic.py::TestRiccati::test_riccati_yn - AssertionError: 
================================================= 1 failed in 0.28s ==================================================
```

### LINKED ISSUE #2078: [SciPy]: DIvergence in handling of do loop var
```fortran
program main
    integer :: n = 2
    do k = 2, n
        print *, k
    end do
    print *, "k after = ", k
end program
```

> GFortran

```console
$ gfortran b.f90  && ./a.out
           2
 k after =            3
```

> LFortran

```console
 lfortran b.f90 --implicit-typing
2
k after =  2
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Continuation of #2079. Fixes #2067, Fixes #2078.

### CODE REVIEW

File: integration_tests/do_loop_04.f90 [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+    do k = 2, n
+        print *, k
+    end do
+
+    ! without --use-loop-variable-after-loop
+
+    if (k /= 2) error stop
+
+    print *, "k after = ", k
+end program
```
Comment (certik) [inline integration_tests/do_loop_04.f90:12]: I would not even test this -- I would reserve the right to change in any way we like.

File: src/libasr/pass/pass_utils.cpp [hunk 2]
```diff
@@ -933,6 +934,10 @@ namespace LCompilers {
                     ASRUtils::EXPR(ASR::make_IntegerBinOp_t(al, loc, a,
                             ASR::binopType::Sub, c, type, nullptr)), nullptr));
 
+                stmt_add_c = ASRUtils::STMT(ASR::make_Assignment_t(al, loc, target,
+                    ASRUtils::EXPR(ASR::make_IntegerBinOp_t(al, loc, a,
+                            ASR::binopType::Add, c, type, nullptr)), nullptr));
+
```
Comment (certik) [inline src/libasr/pass/pass_utils.cpp:940]: You can hide this behind the `use_loop_variable_after_loop` if as well.

File: integration_tests/do_loop_04.f90 (hunk 1 above)
Comment (certik) [reply integration_tests/do_loop_04.f90:12]: Actually, it's good to test that it is not using the new option. So it's good for that. I would put a comment there to update this test if we change what the loop variable is (or even delete the test if it becomes hard to test).

**Comment by Pranavchiku** [comment]
Yes, I will open PRs to apply code review.

**Pull request merged**